packages/optimizer/lib/transformers/ApplyCommonAttributes.js

```diff
@@ -0,0 +1,254 @@
+      this.log.error(error.message);
+      // remove the heights attribute as it's invalid anyway
+      return;
+    }
+    if (!heights.defaultValue) {
+      // remove the sizes attribute as it's invalid anyway
+      return;
+    }
+    this.addHeights(id, heights);
+  }
+
+  addHeights(id, heights) {
+    this.heights.push({
+      id,
+      defaultHeight: heights.defaultValue,
+      // The user agent will pick a width from the sizes attribute, using the first item with a <media-condition> (the part in parentheses) that evaluates to true.
+      // This means, we have to reverse the order the media queries in CSS to emulate this behavior (the last definition has precedence).
+      heights: heights.values.reverse(),
+    });
+  }
+
+  toString() {
+    let result = '';
+    for (const {heights, defaultHeight, id} of this.heights) {
+      const selector = `#${id}:first-child`;
+      result += `${selector}{padding-top:${defaultHeight}}`;
+      for (const height of heights) {
+        result += `@media ${height.media}{${selector}{padding-top:${height.size}}}`;
+      }
+    }
+    return result;
+  }
+}
+
+/**
+ * Stateful attribute transformer. Can only be invoked once per page.
+ */
+class ApplyCommonAttributes {
+  constructor(log) {
+    this.log = log;
+    this.canRemoveBoilerplate = true;
+    // node counter for id generation
+    this.counter = 0;
+    // nodes that have been transformed
+    this.nodes = [];
+    this.attributeTransformations = {
+      media: new MediaTransformer(),
+      sizes: new SizesTransformer(),
+      heights: new HeightsTransformer(),
+    };
+  }
+
+  /**
+   * Applies attribute transformations to the selected node.
+   * @param {Node} node
+   */
+  applyToNode(node) {
+    if (!node.attribs) {
+      return;
+    }
+    for (const [attribute, transformer] of Object.entries(this.attributeTransformations)) {
+      if (hasAttribute(node, attribute)) {
+        try {
+          const id = this.getOrCreateId(node);
+          transformer.transform(node, id);
+          this.nodes.push(node);
+        } catch (e) {
+          this.log.debug(
+            `Cannot remove boilerplate. Failed transforming ${attribute}="${node.attribs[attribute]}".`,
+            e
+          );
+          this.canRemoveBoilerplate = false;
+        }
+      }
+    }
+  }
+
+  /**
+   * Inject custom CSS resulting for attribute transformation.
+   *
+   * @param {Node} head - the head for injecting a styles node if none exists yet
+   * @param {Node|undefined} customStyles - optional existing styles node
+   */
+  applyToCustomStyles(head, customStyles) {
+    const styles = Object.values(this.attributeTransformations).join('');
+    if (!styles) {
+      return;
+    }
+    if (!customStyles) {
+      customStyles = createElement('style', {
+        'amp-custom': '',
+      });
+      appendChild(head, customStyles);
+      insertText(customStyles, '');
+    }
+    customStyles.children[0].data += styles;
```

Comment:
As noted in the description, we'll need to make sure this doesn't go above 75000 bytes.